### PR TITLE
chore: setup borp reporter for switch to node test

### PR DESCRIPTION
**Pull request by dancastillo**
#### Checklist

This PR install `borp` and adds `test/test-reporter.mjs` to run all tap and node:test test suites which will allow to incrementally migrate from tap to node:test. 
This PR is insired from this comment https://github.com/fastify/fastify/pull/5683#issuecomment-2367895455

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,7 +24,7 @@
     "test:validator:integrity": "npm run build:validation && git diff --quiet --ignore-all-space --ignore-blank-lines --ignore-cr-at-eol lib/error-serializer.js && git diff --quiet --ignore-all-space --ignore-blank-lines --ignore-cr-at-eol lib/configValidator.js",
     "test:typescript": "tsc test/types/import.ts --noEmit && tsd",
     "test:watch": "npm run unit -- --watch --coverage-report=none --reporter=terse",
-    "unit": "tap",
+    "unit": "borp --coverage --check-coverage --lines 100 --reporter ./test/test-reporter.mjs",
```
Comment (metcoder95) [inline package.json:27]: Shall we just run the coverage upon `coverage:ci-check-coverage`?

File: test/test-reporter.mjs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+function formatDiagnosticStr (str) {
+  return str.replace(/^(\w+)(\s*\d*)/i, (_, firstWord, rest) => {
+    return firstWord.charAt(0).toUpperCase() + firstWord.slice(1).toLowerCase() + ':' + rest
+  })
+}
+
+async function * reporter (source) {
+  const failed = new Set()
+  const diagnostics = new Set()
+  diagnostics.add('\n\n')
```
Comment (jsumners) [inline test/test-reporter.mjs:26]: Let's leave formatting of output to the display section of the code and only display diagnostic messages when any exist (as we do with failed files).

File: test/test-reporter.mjs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    const whiteText = `\x1b[30m${text}`
+    // Green background with black text
+    return `\x1b[42m${whiteText}\x1b[0m`
+  }
+
+  if (type === 'fail') {
+    const blackText = `\x1b[37m${text}`
+    // Red background with white text
+    return `\x1b[41m${blackText}\x1b[0m`
+  }
```
Comment (jsumners) [inline test/test-reporter.mjs:12]: Can you show a screenshot of what this looks like? I am uncertain black text is going to be visible (to those of us using dark terminal backgrounds).

File: test/test-reporter.mjs
Comment (jsumners) [inline test/test-reporter.mjs:1]: Good work recognizing we need this reporter.

File: package.json [hunk 4]
```diff
@@ -24,7 +24,7 @@
...
     "unit:junit": "tap-mocha-reporter xunit < out.tap > test/junit-testresults.xml",
     "unit:report": "tap --coverage-report=html --coverage-report=cobertura | tee out.tap",
```
Comment (jsumners) [inline package.json:29]: We need to replace all instances of using `tap` to run tests. There's still this one, and at least one more on line 30.

File: test/test-reporter.mjs (hunk 3 above)
Comment (dancastillo) [reply test/test-reporter.mjs]: Dark terminal background:
Screenshot 2024-10-02 at 11 01 10 PM

White terminal background:
Screenshot 2024-10-02 at 11 06 56 PM

**Comment by dancastillo** [comment]
Could this be an issue wiith borp on windows? Looks like macos and ubuntu run without issue
https://github.com/fastify/fastify/actions/runs/11236848075/job/31237887369?pr=5720

**Comment by jsumners** [comment]
It's due to `--reporter=test/test-reporter.mjs` and https://github.com/mcollina/borp/blob/339c6915bc2a782bea1df97e297bff6121c8f801/borp.js#L115-L129

For some reason, it's getting split into `test` and `test-reporter.mjs`. And then `borp` tries to load `node_modules/test/` as a reporter.

**Comment by jsumners** [comment]
Unfortunately, I think we are going to need to extend `borp` to support an rc file in order to get around the various platforms's shell expansions. We should be able to write:

```yaml
# .borp.yaml

reporters:
	- './test/test-reporter.mjs'

files:
	- 'test/**/*.test.js'
	- 'test/**/*.test.mjs'
```

**Comment by jsumners** [comment]
@dancastillo bump borp to 0.18.0 and CI should pass.

File: package.json [hunk 5]
```diff
@@ -10,24 +10,23 @@
     "benchmark": "concurrently -k -s first \"node ./examples/benchmark/simple.js\" \"autocannon -c 100 -d 30 -p 10 localhost:3000/\"",
     "benchmark:parser": "concurrently -k -s first \"node ./examples/benchmark/parser.js\" \"autocannon -c 100 -d 30 -p 10 -i ./examples/benchmark/body.json -H \"content-type:application/jsoff\" -m POST localhost:3000/\"",
     "build:validation": "node build/build-error-serializer.js && node build/build-validation.js",
-    "coverage": "npm run unit -- --coverage-report=html",
-    "coverage:ci": "tap --coverage-report=html --coverage-report=lcov --allow-incomplete-coverage",
-    "coverage:ci-check-coverage": "tap replay",
+    "coverage": "borp --reporter=./test/test-reporter.mjs --coverage --check-coverage --lines 100",
+    "coverage:ci": "nyc --coverage-report=html --coverage-report=lcov borp --reporter=./test/test-reporter.mjs --coverage --check-coverage",
```
Comment (jsumners) [inline package.json:14]: We should be using `c8` for coverage.

File: package.json [hunk 6]
```diff
@@ -10,24 +10,23 @@
...
     "test:report": "npm run lint && npm run unit:report && npm run test:typescript",
     "test:validator:integrity": "npm run build:validation && git diff --quiet --ignore-all-space --ignore-blank-lines --ignore-cr-at-eol lib/error-serializer.js && git diff --quiet --ignore-all-space --ignore-blank-lines --ignore-cr-at-eol lib/configValidator.js",
     "test:typescript": "tsc test/types/import.ts --noEmit && tsd",
     "test:watch": "npm run unit -- --watch --coverage-report=none --reporter=terse",
-    "unit": "tap",
-    "unit:junit": "tap-mocha-reporter xunit < out.tap > test/junit-testresults.xml",
-    "unit:report": "tap --coverage-report=html --coverage-report=cobertura | tee out.tap",
-    "citgm": "tap --jobs=1 --timeout=120"
+    "unit": "borp --reporter=./test/test-reporter.mjs",
+    "unit:report": "nyc --coverage-report=html --reporter=cobertura borp --reporter=./test/test-reporter.mjs --coverage --check-coverage",
```
Comment (jsumners) [inline package.json:28]: `nyc` => `c8`

File: package.json [hunk 7]
```diff
@@ -178,6 +178,7 @@
     "markdownlint-cli2": "^0.13.0",
     "neostandard": "^0.11.3",
     "node-forge": "^1.3.1",
+    "nyc": "^17.1.0",
```
Comment (jsumners) [inline package.json:181]: This should be dropped.

Comment (jsumners) [inline package.json:181]: This should be dropped.

**Review by jsumners — CHANGES_REQUESTED** [review]

File: package.json [hunk 8]
```diff
@@ -10,24 +10,23 @@
...
     "test:report": "npm run lint && npm run unit:report && npm run test:typescript",
     "test:validator:integrity": "npm run build:validation && git diff --quiet --ignore-all-space --ignore-blank-lines --ignore-cr-at-eol lib/error-serializer.js && git diff --quiet --ignore-all-space --ignore-blank-lines --ignore-cr-at-eol lib/configValidator.js",
     "test:typescript": "tsc test/types/import.ts --noEmit && tsd",
     "test:watch": "npm run unit -- --watch --coverage-report=none --reporter=terse",
-    "unit": "tap",
-    "unit:junit": "tap-mocha-reporter xunit < out.tap > test/junit-testresults.xml",
-    "unit:report": "tap --coverage-report=html --coverage-report=cobertura | tee out.tap",
-    "citgm": "tap --jobs=1 --timeout=120"
+    "unit": "borp --reporter=./test/test-reporter.mjs --coverage --check-coverage",
+    "unit:report": "borp --reporter=./test/test-reporter.mjs --coverage > out.test-report",
```
Comment (jsumners) [inline package.json:28]: This is not equivalent. It should be:
```suggestion
    "unit:report": "c8 --reporter html borp --reporter=./test/test-reporter.mjs",
```

File: package.json [hunk 9]
```diff
@@ -10,24 +10,23 @@
     "benchmark": "concurrently -k -s first \"node ./examples/benchmark/simple.js\" \"autocannon -c 100 -d 30 -p 10 localhost:3000/\"",
     "benchmark:parser": "concurrently -k -s first \"node ./examples/benchmark/parser.js\" \"autocannon -c 100 -d 30 -p 10 -i ./examples/benchmark/body.json -H \"content-type:application/jsoff\" -m POST localhost:3000/\"",
     "build:validation": "node build/build-error-serializer.js && node build/build-validation.js",
-    "coverage": "npm run unit -- --coverage-report=html",
-    "coverage:ci": "tap --coverage-report=html --coverage-report=lcov --allow-incomplete-coverage",
-    "coverage:ci-check-coverage": "tap replay",
+    "coverage": "borp --reporter=./test/test-reporter.mjs --coverage --check-coverage --lines 100",
+    "coverage:ci": "borp --reporter=./test/test-reporter.mjs --coverage --check-coverage",
```
Comment (jsumners) [inline package.json:14]: It should be:
```suggestion
    "coverage:ci": "c8 --reporter lcov --reporter html borp --reporter=./test/test-reporter.mjs",
```

File: package.json [hunk 10]
```diff
@@ -10,24 +10,23 @@
     "benchmark": "concurrently -k -s first \"node ./examples/benchmark/simple.js\" \"autocannon -c 100 -d 30 -p 10 localhost:3000/\"",
     "benchmark:parser": "concurrently -k -s first \"node ./examples/benchmark/parser.js\" \"autocannon -c 100 -d 30 -p 10 -i ./examples/benchmark/body.json -H \"content-type:application/jsoff\" -m POST localhost:3000/\"",
     "build:validation": "node build/build-error-serializer.js && node build/build-validation.js",
-    "coverage": "npm run unit -- --coverage-report=html",
-    "coverage:ci": "tap --coverage-report=html --coverage-report=lcov --allow-incomplete-coverage",
-    "coverage:ci-check-coverage": "tap replay",
+    "coverage": "borp --reporter=./test/test-reporter.mjs --coverage --check-coverage --lines 100",
```
Comment (jsumners) [inline package.json:13]:
```suggestion
    "coverage": "c8 --reporter html --check-coverage --100 borp --reporter=./test/test-reporter.mjs",
```

**Review by jsumners — CHANGES_REQUESTED** [review]
Please add a `.borp.yaml` file with the following contents:

```yaml
files:
  - 'test/**/*.test.js'
  - 'test/**/*.test.mjs'
```

We are missing at least one test suite without that configuration.

File: borp.yaml
Comment (jsumners) [inline borp.yaml:1]: The file should be named `.borp.yaml`.

**Review by jsumners — APPROVED** [review]
🎉

**Review by metcoder95 — APPROVED** [review]

File: lib/logger.js [hunk 11]
```diff
@@ -1,10 +1,12 @@
 'use strict'
 
+/* c8 ignore start */
 /**
  * Code imported from `pino-http`
  * Repo: https://github.com/pinojs/pino-http
  * License: MIT (https://raw.githubusercontent.com/pinojs/pino-http/master/LICENSE)
  */
+/* c8 ignore stop */
```
Comment (mcollina) [inline lib/logger.js:9]: I'm pretty sure these should not be needed

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by jsumners** [comment]
I'm going to merge this so work can progress. @dancastillo please determine if the c8 directives Matteo called out are actually needed and follow-up with another PR if they can be removed.

**Pull request merged**